File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Resolution.scala
package org.enso.compiler.core.ir
package expression
package errors

import org.enso.compiler.core.{IR, Identifier}

import java.util.UUID

/** A representation of an error resulting from name resolution.
  *
  * @param originalName the original name that could not be resolved
  * @param reason       the cause of this error
  * @param passData     the pass metadata associated with this node
  * @param diagnostics  compiler diagnostics for this node
  */
sealed case class Resolution(
  originalName: Name,
  reason: Resolution.Reason,
  passData: MetadataStorage      = new MetadataStorage(),
  diagnostics: DiagnosticStorage = DiagnosticStorage()
) extends Error
    with Diagnostic.Kind.Interactive
    with IRKind.Primitive
    with Name
    with LazyId {
  override val name: String = originalName.name

  override def mapExpressions(
    fn: java.util.function.Function[Expression, Expression]
  ): Resolution =
    this

  override def setLocation(
    location: Option[IdentifiedLocation]
  ): Resolution =
    copy(originalName = originalName.setLocation(location))

  /** Creates a copy of `this`.
    *
    * @param originalName the original name that could not be resolved
    * @param reason       the cause of this error
    * @param passData     the pass metadata associated with this node
    * @param diagnostics  compiler diagnostics for this node
    * @param id           the identifier for the new node
    * @return a copy of `this`, updated with the specified values
    */
  def copy(
    originalName: Name             = originalName,
    reason: Resolution.Reason      = reason,
    passData: MetadataStorage      = passData,
    diagnostics: DiagnosticStorage = diagnostics,
    id: UUID @Identifier           = id
  ): Resolution = {
    val res = Resolution(originalName, reason, passData, diagnostics)
    res.id = id
    res
  }

  /** @inheritdoc */
  override def duplicate(
    keepLocations: Boolean   = true,
    keepMetadata: Boolean    = true,
    keepDiagnostics: Boolean = true,
    keepIdentifiers: Boolean = false
  ): Resolution =
    copy(
      originalName = originalName
        .duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
      passData =
        if (keepMetadata) passData.duplicate else new MetadataStorage(),
      diagnostics =
        if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
      id = if (keepIdentifiers) id else null
    )

  /** @inheritdoc */
  override def children: List[IR] = List(originalName)

  /** @inheritdoc */
  override def showCode(indent: Int): String = originalName.showCode(indent)

  /** @inheritdoc */
  override def message(source: (IdentifiedLocation => String)): String =
    reason.explain(originalName)

  /** @inheritdoc */
  override def formattedMessage(
    source: (IdentifiedLocation => String)
  ): String = s"${message(source)}."

  override def diagnosticKeys(): Array[Any] = Array(reason)

  /** @inheritdoc */
  override val location: Option[IdentifiedLocation] = originalName.location
}

object Resolution {

  /** A representation of a symbol resolution error.
    */
  sealed trait Reason {
    def explain(originalName: Name): String
  }

  case object UnresolvedSequenceMacro extends Reason {
    override def explain(originalName: Name): String =
      "No definition for the sequence macro could be found. Try" +
      " importing the default definition from the Standard.Base module"
  }

  /** An error coming from an unknown annotation name.
    */
  case object UnknownAnnotation extends Reason {
    override def explain(originalName: Name): String =
      s"The annotation ${originalName.name} is not defined"
  }

  /** An error coming from a tail call annotation placed in a syntactically
    * incorrect position.
    */
  case object UnexpectedAnnotation extends Reason {
    override def explain(originalName: Name): String =
      s"Unexpected ${originalName.name} annotation. This annotation can " +
      s"only be used with function applications"
  }

  /** An error coming from an unexpected occurence of a polyglot symbol.
    *
    * @param context the description of a context in which the error
    *                happened.
    */
  case class UnexpectedPolyglot(context: String) extends Reason {
    override def explain(originalName: Name): String =
      s"The name ${originalName.name} resolved to a polyglot symbol, " +
      s"but polyglot symbols are not allowed in $context"
  }

  /** An error coming from an unexpected occurence of a constructor.
    *
    * @param context the description of a context in which the error
    *                happened.
    */
  case class UnexpectedConstructor(context: String) extends Reason {
    override def explain(originalName: Name): String =
      s"The name ${originalName.name} resolved to a constructor, " +
      s"but constructors are not allowed in $context"
  }

  /** An error coming from an unexpected occurence of a static method.
    *
    * @param context the description of a context in which the error
    *                happened.
    */
  case class UnexpectedMethod(context: String) extends Reason {
    override def explain(originalName: Name): String =
      s"The name ${originalName.name} resolved to a method, " +
      s"but methods are not allowed in $context"
  }

  /** An error coming from an unexpected occurence of a module.
    *
    * @param context the description of a context in which the error
    *                happened.
    */
  case class UnexpectedModule(context: String) extends Reason {
    override def explain(originalName: Name): String =
      s"The name ${originalName.name} resolved to a module, " +
      s"but modules are not allowed in $context"
  }

  /** An error coming from an unexpected occurence of a type.
    *
    * @param context the description of a context in which the error
    *                happened.
    */
  case class UnexpectedType(context: String) extends Reason {
    override def explain(originalName: Name): String =
      s"The name ${originalName.name} resolved to a type, " +
      s"but types are not allowed in $context"
  }

  /** An error coming from usage of an undefined variable name.
    */
  case object VariableNotInScope extends Reason {
    override def explain(originalName: Name): String =
      s"Variable `${originalName.name}` is not defined"
  }

  /** An error coming from name resolver.
    *
    * @param err the original error.
    */
  case class ResolverError(private val explain: ExplainResolution)
      extends Reason {

    /** Provides a human-readable explanation of the error.
      *
      * @param originalName the original unresolved name.
      * @return a human-readable message.
      */
    override def explain(originalName: Name): String =
      this.explain.explain(originalName)
  }

  trait ExplainResolution {
    def explain(originalName: Name): String
  }

  case class MissingLibraryImportInFQNError(namespace: String) extends Reason {
    override def explain(originalName: Name): String =
      s"Fully qualified name references a library $namespace.${originalName.name} but an import statement for it is missing"
  }

}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/expression/errors/Resolution.java
package org.enso.compiler.core.ir.expression.errors;

import java.util.function.Function;
import org.enso.compiler.core.ir.Diagnostic;
import org.enso.compiler.core.ir.DiagnosticStorage;
import org.enso.compiler.core.ir.Expression;
import org.enso.compiler.core.ir.IRKind;
import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.MetadataStorage;
import org.enso.compiler.core.ir.Name;
import org.enso.compiler.core.ir.expression.Error;
import org.enso.runtime.parser.dsl.GenerateFields;
import org.enso.runtime.parser.dsl.GenerateIR;
import org.enso.runtime.parser.dsl.IRChild;
import org.enso.runtime.parser.dsl.IRField;
import scala.Function1;
import scala.Option;

/** A representation of an erro resulting from name resolution. */
@GenerateIR(
    interfaces = {
      Error.class,
      Diagnostic.Kind.Interactive.class,
      IRKind.Primitive.class,
      Name.class
    })
public final class Resolution extends ResolutionErrorGen {

  /**
   * @param originalName the original name that could not be resolved.
   */
  @GenerateFields
  public Resolution(
      @IRChild Name originalName,
      @IRField Reason reason,
      MetadataStorage passData,
      DiagnosticStorage diagnostics) {
    super(originalName, reason, passData, diagnostics);
  }

  public static Resolution create(Name originalName, Reason reason) {
    return builder().originalName(originalName).reason(reason).build();
  }

  @Override
  public Resolution mapExpressions(Function<Expression, Expression> fn) {
    return this;
  }

  @Override
  public String name() {
    return originalName().name();
  }

  @Override
  public String showCode(int indent) {
    return originalName().showCode(indent);
  }

  @Override
  public String message(Function1<IdentifiedLocation, String> source) {
    return reason().explain(originalName());
  }

  @Override
  public String formattedMessage(Function1<IdentifiedLocation, String> source) {
    return message(source) + ".";
  }

  @Override
  public Object[] diagnosticKeys() {
    return new Object[] {reason()};
  }

  @Override
  public Resolution setLocation(Option<IdentifiedLocation> location) {
    var newOrigName = originalName().setLocation(location);
    return builder(this).originalName(newOrigName).build();
  }

  @Override
  public IdentifiedLocation identifiedLocation() {
    return originalName().identifiedLocation();
  }

  @Override
  public Option<IdentifiedLocation> location() {
    return Option.apply(identifiedLocation());
  }

  sealed interface Reason {
    String explain(Name originalName);
  }

  public static final class UnknownAnnotation implements Reason {
    private UnknownAnnotation() {}

    public static final UnknownAnnotation INSTANCE = new UnknownAnnotation();

    @Override
    public String explain(Name originalName) {
      return "The annotation " + originalName.name() + " is not defined";
    }
  }

  /** An error coming from a tail call annotation placed in a syntactically incorrect position. */
  public static final class UnexpectedAnnotation implements Reason {
    private UnexpectedAnnotation() {}

    public static final UnexpectedAnnotation INSTANCE = new UnexpectedAnnotation();

    @Override
    public String explain(Name originalName) {
      return "Unexpected "
          + originalName.name()
          + " annotation. This annotation can "
          + "only be used with function applications";
    }
  }

  /**
   * An error coming from an unexpected occurence of a polyglot symbol.
   *
   * @param context the description of a context in which the error happened.
   */
  public record UnexpectedPolyglot(String context) implements Reason {
    @Override
    public String explain(Name originalName) {
      return "The name "
          + originalName.name()
          + " resolved to a polyglot symbol, "
          + "but polyglot symbols are not allowed in "
          + context;
    }
  }

  /**
   * An error coming from an unexpected occurence of a constructor.
   *
   * @param context the description of a context in which the error happened.
   */
  public record UnexpectedConstructor(String context) implements Reason {
    @Override
    public String explain(Name originalName) {
      return "The name "
          + originalName.name()
          + " resolved to a constructor, "
          + "but constructors are not allowed in "
          + context;
    }
  }

  /**
   * An error coming from an unexpected occurence of a static method.
   *
   * @param context the description of a context in which the error happened.
   */
  public record UnexpectedMethod(String context) implements Reason {
    @Override
    public String explain(Name originalName) {
      return "The name "
          + originalName.name()
          + " resolved to a method, "
          + "but methods are not allowed in "
          + context;
    }
  }

  /**
   * An error coming from an unexpected occurence of a module.
   *
   * @param context the description of a context in which the error happened.
   */
  public record UnexpectedModule(String context) implements Reason {
    @Override
    public String explain(Name originalName) {
      return "The name "
          + originalName.name()
          + " resolved to a module, "
          + "but modules are not allowed in "
          + context;
    }
  }

  /**
   * An error when a project-private entity (module, type, method) is used from a different project.
   *
   * @param callerProject Name of the project of caller.
   * @param calleeProject Name of the project of callee.
   */
  public record PrivateEntity(String callerProject, String calleeProject) implements Reason {
    @Override
    public String explain(Name originalName) {
      return "Project-private entity '"
          + originalName.name()
          + "' in project '"
          + calleeProject
          + "' cannot be used from project '"
          + callerProject
          + "'";
    }
  }

  /** An error coming from name resolver. */
  public record ResolverError(ExplainResolution explain) implements Reason {

    @Override
    public String explain(Name originalName) {
      return explain.explain(originalName);
    }
  }

  public interface ExplainResolution {
    String explain(Name originalName);
  }

  public record MissingLibraryImportInFQNError(String namespace) implements Reason {

    @Override
    public String explain(Name originalName) {
      return "Fully qualified name references a library "
          + namespace
          + "."
          + originalName.name()
          + " but an import statement for it is missing";
    }
  }
}
